### PR TITLE
Enable custom cert for app

### PR DESCRIPTION
**Pull request by BardOve**
To enable support for custom certificates in skiperator applications, we add a field customCertificate to the application spec. This  points to a tls secret in istio-gateways namespace to use in gateway resource.

### CODE REVIEW

File: controllers/application/certificate.go [hunk 1]
```diff
@@ -61,23 +61,37 @@ func (r *ApplicationReconciler) reconcileCertificate(ctx context.Context, applic
...
+				return nil
+			})
+		} else {
+			secret, err := util.GetSecret(r.GetClient(), ctx, types.NamespacedName{ Namespace: "istio-gateways", Name: application.Spec.CustomCertificateSecret })
+			if err != nil {
+				fmt.Errorf("Failed to get secret %s", application.Spec.CustomCertificateSecret)
+				r.SetControllerError(ctx, application, controllerName, err)
+				return util.DoNotRequeue()
 			}
+			if secret.Type != "kubernetes.io/tls" {
```
Comment (evenh) [inline controllers/application/certificate.go:88]: Required import:
```
corev1 "k8s.io/api/core/v1"
```

```suggestion
			if secret.Type != corev1.SecretTypeTLS {
```

**Review by evenh — CHANGES_REQUESTED** [review]
After looking through the code and giving it some thought I think there's two major unhandled issues.

### Routing
We'll need to handle the `controllers/routing` package as well. 

### Hostnames
An application can have multiple hostnames. Right now we're just overriding with whatever custom certificate secret that's passed without considering whether this is for an internal or an external ingress, or if the hostname in the gateway matches whatever is present in the certificate secret.

I suggest that we team up and fix those issues together :-)

File: api/v1alpha1/application_types.go [hunk 2]
```diff
@@ -64,6 +65,8 @@ type ApplicationSpec struct {
 	// HTTP and HTTPS.
 	//
 	// Ingresses must be lowercase, contain no spaces, be a non-empty string, and have a hostname/domain separated by a period
+	// They can optionally be suffixed with a plus and name of a custom TLS secret located in the istio-gateways namespace.
```
Comment (martinhny) [inline api/v1alpha1/application_types.go:68]: I think we should have a CustomCerts type or turn this one into an interface which accepts both an object and string, although this is a dirty solution as well. Or will this make reconciling very difficult?

File: config/crd/skiperator.kartverket.no_applications.yaml [hunk 3]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.15.0
```
[the remark this answers is not in the document]
Comment (evenh) [reply config/crd/skiperator.kartverket.no_applications.yaml]: Nope, a side effect of looking up binaries in `$PATH` before `$GOBIN`. Fixed in Makefile :)

File: api/v1alpha1/hosts.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		h = Host{Hostname: results[0], CustomCertificateSecret: &secret}
+	// More than one '+' characters present
+	default:
+		return nil, fmt.Errorf("%s: not valid, contains multiple '%s' characters", hostname, hostnameSecretSeparator)
+	}
+
+	// Verify that the hostname is an actual valid DNS name.
+	if err := domain.Check(h.Hostname); err != nil {
+		return nil, fmt.Errorf("%s: failed validation: %w", h.Hostname, err)
+	}
```
[the remark this answers is not in the document]
Comment (evenh) [reply api/v1alpha1/hosts.go:47]: My thoughts exactly as well. We might be in a situation where applications are deployed before DNS are provisioned.

File: controllers/application/controller.go [hunk 5]
```diff
@@ -257,10 +259,16 @@ func (r *ApplicationReconciler) isCrdPresent(ctx context.Context, name string) b
...
+	var err error
+	hosts, err := application.Spec.Hosts()
+	if err != nil {
+		return err
+	}
+
+	// TODO: Remove/rewrite?
+	for _, h := range hosts {
+		if !hostMatchExpression.MatchString(h.Hostname) {
+			errMessage := fmt.Sprintf("ingress with value '%s' was not valid. ingress must be lower case, contain no spaces, be a non-empty string, and have a hostname/domain separated by a period", h.Hostname)
```
[the remark this answers is not in the document]
Comment (evenh) [reply controllers/application/controller.go:271]: With the time available before the holidays I couldn't get the tests to pass when rewriting this section. Happy to team up in August.

**Review by evenh — APPROVED** [review]

**Pull request merged**